enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
+Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
+new instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneMachineSet
+
+As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
+`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
+the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
+instances. This will prevent the garbage collector from removing the Control Plane Machines when the
+`ControlPlaneMachineSet` is deleted.
+
+If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.
+
+#### Installing a ControlPlaneMachineSet within an existing cluster
+
+When adding a ControlPlaneMachineSet to a existing cluster, the end user will need to define the
+`ControlPlaneMachineSet` resource by copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
+field within the ControlPlaneMachineSet spec.
+
+To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
+that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.
+We will also validate that the failure domains align with those within the cluster already, this will prevent users
+from accidentally migrating their Control Plane Machines from multiple availability zones to a single zone.
+
+If no Control Plane Machines exist, or they are in a non-Running state, the operator will report degraded until this
+issue is resolved. This creates a dependency for the `ControlPlaneMachineSet` operator on the Machine API. It will be
+required to run at a higher run-level than Machine API.
+
+In UPI or misconfigured clusters, a user adding a `ControlPlaneMachineSet` will result in a degraded cluster.
+Users will need to remove the invalid `ControlPlaneMachineSet` resource to restore their cluster to a healthy state.
+
+We do not recommend that UPI users attempt to adopt their control plane instances into Machine API due to the high
```

Comment:
I do not agree with this sentiment. This will just add another tick to making UPI a lower-class cluster. I do not think that the likelihood is high that the user cannot create an accurate configuration. I would expect in most cases that the configuration would be pretty close to what the IPI would have created already.

Reply:
I'm going to reword this section where it talks about UPI, you're right, we shouldn't be making this distinction, I just want to make sure it's clear that users must ensure they have "Running" Control Plane Machines before they create the CPMS